cirq/study/resolver.py

```diff
@@ -154,6 +155,13 @@ def value_of(self,
         # No known way to resolve this variable, return unchanged.
         return value
 
+    def _resolve_parameters_(self, param_resolver: 'ParamResolver'
+                            ) -> 'ParamResolver':
+        return ParamResolver({
+            k: protocols.resolve_parameters(v, param_resolver)
+            for k, v in self.param_dict.items()
+        })
```

Comment:
In order to avoid surprises, composition should be [associative](https://en.wikipedia.org/wiki/Associative_property), but this implementation isn't. Consider

```python
pr1 = ParamResolver({a: b})
pr2 = ParamResolver({})
pr3 = ParamResolver({b: 10})
```

Since `self` (i.e. LHS) controls what keys go into the result, we see that pr2+pr3 is empty and therefore pr1+(pr2+pr3) is pr1. OTOH, pr1+pr2 is just pr1 so (pr1+pr2)+pr3 is `ParamResolver({a: 10})`.

Reply:
I suspect that it is not feasible to achieve this without significant changes to the core behavior of `ParamResolver`. My basis for this is that `cirq.resolve_parameters` attempts to fully resolve each parameter using the given resolver, but only succeeds if the resolution proceeds in alphabetical order:
```
a, b, c, d = [sympy.Symbol(x) for x in ['a', 'b', 'c', 'd']]

# Multiple resolution steps are taken...
assert cirq.resolve_parameters(a, {a: b, b: c}) == c

# ...but only in alphabetical order.
assert cirq.resolve_parameters(c, {c: b, b: a}) == b
```
This is not a Cirq issue - it's just how `sympy.Basic.subs` behaves. From `help(sympy.Basic.subs)`:
> In order to obtain a canonical result, unordered iterables are sorted by count_op length, number of arguments and by the default_sort_key to break any ties.

Example:
```
assert c.subs({c: b, b: a}) == b
```
I tried one-step parameter resolution, but that causes issues when reassigning values:
```
r1 = cirq.ParamResolver({a: b})
r2 = cirq.ParamResolver({c: d, d: 1})
r3 = cirq.ParamResolver({d: 2})

r12 = cirq.resolve_parameters(r1, r2)    # {a: b, c: 1, d: 1}
r23 = cirq.resolve_parameters(r2, r3)    # {c: 2, d: 2}
r12_3 = cirq.resolve_parameters(r12, r3)  # {a: b, c: 1, d: 2} 
r1_23 = cirq.resolve_parameters(r1, r23)  # {a: b, c: 2, d: 2}
```
-----
This could probably be resolved by enforcing one-step parameter resolution in `ParamResolver` with `simultaneous=True` in the `subs` call, but that's a breaking change.

Reply:
Opened #3544 to track this.

Reply:
Ah, this is a fairly subtle problem. Thanks for identifying it and for filing the issue!

I'm still wondering whether we could change this function here in a way that would make it easy for us to make resolver composition associative if/when the resolution mechanism is made one-step.

One way to avoid the problem in your example would be to require that there be no conflicts between resolvers being composed. Specifically, we could raise an exception if the two resolvers both mapped the same key to a different value. WDYT?

Reply:
There are two issues that break associativity in the current version: lack of implicit maps (k: k for all unlisted keys) and mixing of resolution depth (`self.value_of` vs. `self.param_dict.get`). In a one-step resolution world, the following should work:
```
def _resolve_parameters_(self, param_resolver):
  new_dict = {k: k for k in param_resolver}
  new_dict.update({k: self.value_of(k) for k in self})
  new_dict.update({k: param_resolver.value_of(v) for k, v in new_dict.items()})
  return ParamResolver(new_dict)
```
(This may also work for full resolution, but I'm not as certain on that point.)

In practice:
```
a, b, c, d = [sympy.Symbol(x) for x in ['a', 'b', 'c', 'd']]
r1 = cirq.ParamResolver({a: b})
r2 = cirq.ParamResolver({c: d, d: 1})
r3 = cirq.ParamResolver({d: 2})

r12 = cirq.resolve_parameters(r1, r2)    # {a: b, c: d, d: 1} - no "chain resolution" from the new resolver
r23 = cirq.resolve_parameters(r2, r3)    # {c: 2, d: 1}
r12_3 = cirq.resolve_parameters(r12, r3)  # {a: b, c: 2, d: 1} 
r1_23 = cirq.resolve_parameters(r1, r23)  # {a: b, c: 2, d: 1}
```
Despite "losing" the `d: 2` mapping from r3, I believe that this gives the correct behavior, since r2 removes any "initial-d" parameters from the circuit. The only parameters that care about the new d-mapping are those that are "currently" d when r3 is applied (e.g. "initial-c" parameters):
```
q0, q1, q2, q3 = cirq.LineQubit.range(4)
c0 = cirq.Circuit(cirq.X(q0) ** a, cirq.X(q1) ** b, cirq.X(q2) ** c, cirq.X(q3) ** d)

c1 = resolve_parameters(c0, r1)  # X(q0)^b, X(q1)^b, X(q2)^c, X(q3)^d
c2 = resolve_parameters(c1, r2)  # X(q0)^b, X(q1)^b, X(q2)^d, X(q3)^1
c3 = resolve_parameters(c2, r3)  # X(q0)^b, X(q1)^b, X(q2)^2, X(q3)^1

c123 = cirq.resolve_parameters(c0, r12_3)
assert c123 == c3
```

Reply:
> (This may also work for full resolution, but I'm not as certain on that point.)

This is only partially true. While the full-resolution behavior of the two resolvers will be the same, their internal mappings may differ, as seen below:
```
r1 = cirq.ParamResolver({a: b})
r2 = cirq.ParamResolver({x: y})
r3 = cirq.ParamResolver({b: x})

r12 = cirq.resolve_parameters(r1, r2)    # {a: b, x: y}
r12_3 = cirq.resolve_parameters(r12, r3)  # {a: x, x: y, b: x} - maps all to y, but transitively
r23 = cirq.resolve_parameters(r2, r3)    # {b: x, x: y} - loss of ordering occurs here
r1_23 = cirq.resolve_parameters(r1, r23)  # {a: y, x: y, b: y}
```
The fix is to "flatten" the resulting ParamResolver post-composition. This has the added perk of simplifying the internal mapping, which speeds up future parameter resolution:
```
def flatten(self):
  # resolving the empty dict against a PR flattens that PR
  return ParamResolver()._resolve_parameters_(self, recursive=True)
```

Reply:
Nice analysis. I think this should work. In particular, I think you're right that we don't need to raise an exception when there is a conflict as long as we choose consistently which resolver wins (above we choose RHS which is fine). After all, commutativity is *not* a requirement (just as it is not in case of other forms of composition).

I think this is tricky enough that we should really have a number of unit tests that exercise interesting scenarios (see other comment), each in both "once" and "recursive" modes (as long as both modes are possible; some scenarios will probably fail in the "recursive" mode, e.g. a: a+1). The cases we considered above are probably a good starting point.